### PR TITLE
feat: rework binding between Servers and MetalMachines

### PR DESCRIPTION
**Pull request by smira**
New resource was introduced: `ServerBinding` which matches 1:1 to the
Server (same name), and has reference to `MetalMachine` and (optionally)
`ServerClass` server was picked from.

Signed-off-by: Andrey Smirnov <smirnov.andrey@gmail.com>

### CODE REVIEW

File: app/cluster-api-provider-sidero/api/v1alpha3/serverbinding_types.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+type ServerBindingState struct {
+	// Ready is true when matching server is found.
+	// +optional
+	Ready bool `json:"ready"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:printcolumn:name="Ready",type="string",JSONPath=".status.ready",description="ServerBinding ready status"
+// +kubebuilder:printcolumn:name="MetalMachine",type="string",priority=1,JSONPath=".spec.metalMachineRef.name",description="Metal Machine"
```
Comment (smira) [inline app/cluster-api-provider-sidero/api/v1alpha3/serverbinding_types.go:31]: server, server class, metal machine, cluster

**Review by rsmitty — DISMISSED** [review]

**Comment by smira** [comment]
I tested the upgrade via `clusterctl`, worked great and smooth. (From latest Sidero release to this PR version). ServerBindings were created to match the current allocation and plus the Server's allocated status didn't change during the upgrade.

**Pull request merged**